### PR TITLE
Feature/task 4173 add edit entity

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Add or edit location or person screen

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4173

## Screenshots
Simulator Screen Shot - iPhone 12 Pro - 2020-12-14 at 17 55 15
Simulator Screen Shot - iPhone 12 Pro - 2020-12-15 at 16 49 26

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift [hunk 1]
```diff
@@ -5,16 +5,121 @@
...
+
+	// MARK: - Internal
 
 	enum Mode {
 		case add(DiaryDay, DiaryEntryType)
 		case edit(DiaryEntry)
 	}
 
+	let mode: Mode
+	// is store maybe enougth?
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift:42]: Comment still valid?

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift [hunk 2]
```diff
@@ -5,16 +5,121 @@
...
 
-struct DiaryAddAndEditEntryViewModel {
+class DiaryAddAndEditEntryViewModel {
+
+	// MARK: - Init
+
+	init(
+		mode: Mode,
+		diaryService: DiaryService,
+		dismiss: @escaping () -> Void
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift:15]: What is the benefit of routing the user interaction through the view model?
The closure can be just part of the ViewController.

Also from the calling point of view. Calling "viewModel.dismiss()".
What is dismissed from the viewmodel? Some data? It sounds a little bit strange.

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/Cells/DiaryDayEntryTableViewCell.xib [hunk 3]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/Cells/DiaryDayEntryTableViewCell.xib:1]: OMG 😳 I need to open IB? Hopefully by MacBook does not explode 🤪

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+			return viewModel.entriesOfSelectedType.count
+		default:
+			fatalError("Invalid section")
+		}
+	}
+
+	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+		switch indexPath.section {
+		case 0:
+			return addCell(forRowAt: indexPath)
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift:77]: It reads like a cell is added somewhere. 

```suggestion
			return makeAddCell(forRowAt: indexPath)
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+			viewModel.toggleSelection(of: viewModel.entriesOfSelectedType[indexPath.row])
+		default:
+			fatalError("Invalid section")
+		}
+	}
+
+	// MARK: - Private
+
+	private let viewModel: DiaryDayViewModel
+	private let onAddEntryCellTap: (DiaryDay, DiaryEntryType) -> Void
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift:101]: I like this a lot more. Without routing the user interaction through the view model  👍.

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+	private var subscriptions = [AnyCancellable]()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift:103]: Should subscriptions be deleted on deinit? 🤔

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+	init(day: DiaryDay, store: DiaryStoring) {
+		self.day = day
+		self.store = store
+
+		store.diaryDaysPublisher
+			.sink { [weak self] days in
+				guard let day = days.first(where: { $0.dateString == day.dateString }) else { return }
+
+				self?.day = day
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift:20]: can the day switch while I have opened a specific day?

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	@Published var selectedEntryType: DiaryEntryType = .contactPerson
+
+	var entriesOfSelectedType: [DiaryEntry] {
+		day.entries.filter {
+			$0.type == selectedEntryType
+		}
+	}
+
+	func toggleSelection(of entry: DiaryEntry) {
+		entry.isSelected ? deselect(entry: entry) : select(entry: entry)
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift:36]: Should `entry.isSelected` not toggle after selecting/deselecting the stuff?

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/EditEntries/DiaryEditEntriesViewController.swift [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+				self.shouldReload = true
+			})
+		}
+	}
+
+	// MARK: - Private
+
+	private let viewModel: DiaryEditEntriesViewModel
+	private let onCellSelection: (DiaryEntry) -> Void
+	private let onDismiss: () -> Void
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/EditEntries/DiaryEditEntriesViewController.swift:109]: Still needed?

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift (hunk 1 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift]: Done

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift (hunk 2 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift]: Reworked it

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift (hunk 6 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift]: The are cancelled automatically on deinitalization

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift (hunk 4 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift]: This is from a different branch, the diff was not correct here in draft mode.

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift (hunk 7 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift]: Not the day but it's entries can change

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/Cells/DiaryDayEntryTableViewCell.swift [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+		case .location(let location):
+			checkboxImageView.image = location.isSelected ? UIImage(named: "Diary_Checkmark_Selected") : UIImage(named: "Diary_Checkmark_Unselected")
+			label.text = location.name
+		}
+
+		if entry.isSelected {
+			accessibilityTraits = [.button, .selected]
+		} else {
+			accessibilityTraits = [.button]
+		}
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/Cells/DiaryDayEntryTableViewCell.swift]: This was from a different pull request and the diff wasn't right yet in draft mode, but I'll comment anyways: We don't use the selection state of the cell, that's why we have to set in manually.

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift (hunk 8 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewModel.swift]: That is triggered by the store in the model layer and then comes back through the publisher in the correct selected/deselected state.

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/EditEntries/DiaryEditEntriesViewController.swift (hunk 9 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/EditEntries/DiaryEditEntriesViewController.swift]: Yep, its used from the bar button item

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	private func updateForSelectedEntryType() {
+		tableView.reloadData()
+
+		if viewModel.entriesOfSelectedType.isEmpty {
+			tableView.backgroundView = DiaryDayEmptyView(entryType: viewModel.selectedEntryType)
+		} else {
+			tableView.backgroundView = nil
+		}
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/Day/DiaryDayViewController.swift]: I like it, will fix it on another branch 😊

**Review by nickguendling — APPROVED** [review]

**Review by janetback — CHANGES_REQUESTED** [review]
Hi, I have added a missing full stop (sorry, this was also missing in Figma) - > if ok with you please resubmit and I will approve.

**Review by janetback — APPROVED** [review]
UA reviewed.

**Pull request merged**